lib/cache.ml

```diff
@@ -0,0 +1,68 @@
+open Bos
+open Rresult
+open Astring
+
+let pp_header = Fmt.(styled `Blue (const string "==> "))
+
+let valid_remote_name = String.map (function '/' | '@' | ':' -> '-' | x -> x)
+
+let git_remote_exists_or_create ~remotes ~remote =
+  let remote_name = valid_remote_name remote in
+  match String.Set.mem remote_name remotes with
+  | true -> Ok ()
+  | false -> Exec.git_remote_add ~remote_url:remote ~remote_name
+
+let git_branch_exists_or_create ~output_dir ~remotes ~branch ~remote ~tag cache_dir =
+  let remote_name = valid_remote_name remote in
+  OS.Dir.with_current cache_dir
+    (fun () ->
+      OS.Path.exists Fpath.(v ".git" / "refs" / "heads" / branch) >>= function
```

Comment:
this will not work with `packed-refs` -- e.g. after you run a `git gc`. Usually it's really unsafe to look in `.git/` directly as the Git tool will often optimise things behind your back.